File: LA Traffic Prediction/GTNN_MATLAB/GTNN.m
classdef GTNN < handle
    % Rectangle class with width and height properties and area/perimeter methods
    properties
        ANET = handle([]);
        GNET = handle([]);
        memupdaters;
        numnodes, numfeats, numOutputFeats;

        adj, adj_pre, adjs, adjs_u;
        dcd_adjs, dcd_adj, dcdadj_pre, dadj_dadjpre;
    end
    
    methods
        function obj = GTNN(numnodes, numfeats, numOutputFeats)
            % general params
            obj.numnodes = numnodes; obj.numfeats = numfeats; obj.numOutputFeats = numOutputFeats;

            % init A-NET & G-NET
            obj.init_ANET([numfeats*2 4 1], ["relu" "sigmoid" "leaky_relu"]);
            obj.init_GNET([numfeats numOutputFeats], ["relu" "N/A"]);
            obj.init_memupdaters([2 4 4], ["relu" "sigmoid" "LSTM_custom"]);       % init memupdaters

        end

        function init_ANET(obj, layer_sizes, activations)
            obj.ANET = GTNN_NN(1, layer_sizes, activations);
            for k = obj.ANET.numLayers:-1:1         % dec. all the ANET learning rates for fair training
                obj.ANET.layers(k).lrate = obj.ANET.layers(k).lrate * 1 / obj.numnodes^(2);
            end
        end
        function init_GNET(obj, laysizes, activations)
            obj.GNET = GTNN_NN(obj.numnodes, laysizes, activations);
        end
        function init_memupdaters(obj, laysizes, activations)
            % init mem updaters
            obj.memupdaters = LSTM.empty(obj.numfeats,0);
            for f = 1:obj.numfeats
                obj.memupdaters(f) = LSTM(obj.numnodes, laysizes, activations);
            end
        end
        
        function forward(obj, state_in)
            % adj = ANET.forward(state_in)
            for n = 1:obj.numnodes
                for c = 1:obj.numnodes
                    obj.adj_pre(n,c) = obj.ANET.forward_gat( [ state_in(c,:) state_in(n,:) ] );
                end
                obj.adj(n,:) = softmax_(obj.adj_pre(n,:));
            end

            % adjs = adj * state_in
            obj.adjs = obj.adj * state_in;
            for f= 1:obj.numfeats
                % adjs_u(:,f) = memupdaters(f).ht
                obj.memupdaters(f).forward(obj.adjs(:,f));
                obj.adjs_u(:,f) = obj.memupdaters(f).ht;
            end

            % graph pred = GNET.forward(adjs_u)
            obj.GNET.forward(obj.adjs_u);
        end

        function update_allmem(obj)
            for f = 1:obj.numfeats
                obj.memupdaters(f).update_mem();
            end
        end      
        function reset_allmem(obj)
            for f = 1:obj.numfeats
                obj.memupdaters(f).reset_mem();
            end
        end
        
        function avgcost = backward(obj, state_in, strue)
            avgcost = obj.GNET.backward(strue);
            for f = 1:obj.numfeats
                obj.memupdaters(f).backward_using_deriv(obj.GNET.layers(1).dcda(:,f));
            end
            for f = 1:obj.numfeats
                obj.dcd_adjs(:,f) = obj.memupdaters(f).feedinNN.layers(1).dcda(:,2);
            end
            obj.dcd_adj = obj.dcd_adjs * state_in';

            for n = 1:obj.numnodes
                obj.dadj_dadjpre = dsoftmax(obj.adj(n,:));
                obj.dcdadj_pre(n,:) = obj.dcd_adj(n,:) * obj.dadj_dadjpre;
                for c = 1:obj.numnodes
                    obj.ANET.backward_gat([state_in(c,:) state_in(n,:)], obj.adj_pre(n,c), obj.dcdadj_pre(n,c));
                end
            end
        end

        function avgcost = pred_tplus(obj, sin, strue, nsteps)
            for tp = 1:nsteps
                obj.forward(sin);
                sin = obj.GNET.layers(end).a;
                obj.update_allmem();
            end
            disp([sin strue]);
            avgcost = mean(obj.GNET.layers(end).a(:) - strue(:));
            disp(avgcost);
        end

        function test(obj, xdata, nsteps)
            numB = floor ( height(xdata) / obj.numnodes );
            numSeq = floor ( numB / nsteps) ;
            for k = 1:numSeq - 1
                obj.pred_tplus(getbatch(k*nsteps, obj.numnodes, xdata), getbatch((k+1)*nsteps, obj.numnodes, xdata), nsteps);
            end
        end

        function train(obj, nepochs, xdata)
            numB = floor ( height(xdata) / obj.numnodes ) ;
            for k = 1:nepochs
                obj.reset_allmem();
                for b = 1:numB-1
                    s_in = xdata((b-1)*obj.numnodes + 1: (b-1)*obj.numnodes + obj.numnodes, :);
                    s_true = xdata((b)*obj.numnodes + 1: (b)*obj.numnodes + obj.numnodes, :);

                    obj.forward( s_in );
                    avgcost = obj.backward( s_in, s_true );
                    if randi(500) == 1
                        disp([ obj.GNET.layers(end).a s_true ]);
                        disp(avgcost);
                    end
                    obj.update_allmem();
                end
            end
        end


        
    end
end

File: LA Traffic Prediction/GTNN_C++_METRLA_TrafficPred/GTNN_C++/GTNN/GTNN.cpp
 
//+++++++++++++++++++++++++++++++++++++++++++++++
// THIS CODE IS PROPERTY OF AYUSH SAHA (2/9/2023)
// Undergraduate student at UC DAVIS, College of Engineering
// DO NOT DISTRIBUTE
//+++++++++++++++++++++++++++++++++++++++++++++++


#include "GTNN.h"

#include<iostream>
#include<unordered_map>
#include<random>

//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// INIT

// main init variables
GTNN::GTNN(int numnodes, int numfeats) {
	this->numnodes = numnodes;
	this->numfeats = numfeats;
	init_matrices();
}
// INITIALIZE matrix sizes
void GTNN::init_matrices() {
	resizeMatrix(adj_pre, numnodes, numnodes);
	resizeMatrix(adj, numnodes, numnodes);
	resizeMatrix(adjs, numnodes, numfeats);
	resizeMatrix(adjs_u, numnodes, numfeats);

	resizeMatrix(dcd_adj, numnodes, numnodes);
	resizeMatrix(dcd_adjs, numnodes, numfeats);
	resizeMatrix(dadj_dadjpre, numnodes, numnodes);
	resizeMatrix(dcdadj_pre, numnodes, numnodes);

}
// rangomize A-NET & G-NET W's and B's
void GTNN::randomize_WBs(std::uniform_real_distribution<float>& dist, std::default_random_engine& gen) {
	anet->randomize_WBs(dist, gen);
	gnet->randomize_WBs(dist, gen);
}
// initialize A-NET
void GTNN::init_anet(int numLayers, int layer_sizes_[], std::string layer_activations[]) {
	int numfeats_anet = numfeats * 2;
	anet = new NN_GTNN( numfeats_anet , 1, 1, numLayers, layer_sizes_, layer_activations);
}
// init G-NET
void GTNN::init_gnet(int numLayers, int layer_sizes_[], std::string layer_activations[]) {
	gnet = new NN_GTNN(numfeats, numfeats, numnodes, numLayers, layer_sizes_, layer_activations);
}
// initialize ALL memupdaters
void GTNN::init_memupdaters(int numLayers, int layer_sizes[], std::string layer_activations[], \
	std::uniform_real_distribution<float>& dist, std::default_random_engine& gen) {

	memupdaters = (LSTM**)malloc(numfeats * sizeof(LSTM*));
	for (int f = 0; f < numfeats; f++) {
		memupdaters[f] = new LSTM(numnodes);
		memupdaters[f]->init_feedinNN(numLayers, layer_sizes, layer_activations);
		memupdaters[f]->randomize_WBs(dist, gen);
		memupdaters[f]->set_gdatacol(f);
	}
}

//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// FORWARD PROP

void GTNN::forward(float**& state_in) {
	for (int n = 0; n < numnodes; n++) {
		for (int c = 0; c < numnodes; c++) {
			// attention = A-NET ( state_in[c] | state_in[n] )
			adj_pre[n][c] = anet->predict(state_in[c], state_in[n]);
		}
		// attentions = softmax ( attentions )
		apply_softmax(adj_pre[n], adj[n], numnodes);
	}
	// adjs = adj * state_in
	matmul(adj, state_in, adjs, numnodes, numnodes, numfeats);

	// adjs_updated[:, feat] = memupdaters[f]->forward ( adjs[:,feat] )
	for (int f = 0; f < numfeats; f++) {
		memupdaters[f]->forward(adjs, adjs_u);
	}

	// G-NET forward props on adjs matrix
	gnet->forward(adjs_u);
}

//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// BACK PROP

float GTNN::backward(float**& state_in, float**& strue) {
	float graphpred_cost = gnet->backward(strue);		// cost for graph-state predictions

	float**& dcd_adjsu = gnet->get_layer0_dcda();
	for (int f = 0; f < numfeats; f++) {
		// backprop memupdaters[f]
		memupdaters[f]->backward_using_derivative(dcd_adjsu);
		// dcd_adjs[:,feat] = memupdaters[f]->dcdxt[:,0]
		memupdaters[f]->get_dcdxt(dcd_adjs);
	}

	// dcd_adj = dcdadjs * state_in
	mat_times_mat_T(dcd_adjs, state_in, dcd_adj, numnodes, numfeats, numnodes);

	// float attention_cost = get_avg_value(dcd_adj, numnodes, numnodes);		// cost for attention predictions
	float learnrate_scale = (1 / std::pow(numnodes, 2));
	for (int n = 0; n < numnodes; n++) {

		de_softmax(adj[n], dadj_dadjpre, numnodes);
		vect_times_mat(dcd_adj[n], dadj_dadjpre, dcdadj_pre[n], numnodes, numnodes);

		for (int c = 0; c < numnodes; c++) {
			anet->backward(state_in[c], state_in[n], dcdadj_pre[n][c], learnrate_scale);
		}
	}
	return graphpred_cost;
}

//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// TRAIN/TEST ROUTINES

void GTNN::show_preds(float**& strue, int t) {
	float**& preds = gnet->get_preds();

	std::cout << "\nPREDICTIONS @ t=" << std::to_string(t) << "\n";
	for (int j = 0; j < numfeats; j++) {
		std::cout << "feat" + std::to_string(j) + "_in" << " " << \
			"feat" + std::to_string(j) + "_true" << "\t";
	} std::cout << "\n";
	for (int i = 0; i < numnodes; i++) {
		for (int j = 0; j < numfeats; j++) {
			std::cout << preds[i][j] << " " << strue[i][j] << "\t";
		}
		std::cout << "\n";
	}
	std::cout << "\n\n";
	std::cout << "MAE: " << get_avg_diff(preds, strue, numnodes, numfeats) << "\n";
}

void GTNN::reset_all_mem() {
	for (int f = 0; f < numfeats; f++) {
		memupdaters[f]->reset_mem();
	}
}
void GTNN::update_all_mem() {
	for (int f = 0; f < numfeats; f++) {
		memupdaters[f]->update_mem();
	}
}
void GTNN::update_all_mem(float**& dcdht) {
	for (int f = 0; f < numfeats; f++) {
		memupdaters[f]->update_mem(dcdht);
	}
}

void GTNN::train(int num_epochs) {
	for (int p = 0; p < num_epochs; p++) {
		reset_all_mem();
		for (int t = 0; t < ntimesteps-1; t++) {
			// get state_in and state_true matrices
			float**& sin = dstore->get_g_state(t);
			float**& strue = dstore->get_g_state(t+1);

			// forward prop./back prop
			forward(sin);		// forward prop.
			float MAE = backward(sin, strue);		// back prop.

			// update_all_mem();

			// display predictions
			if (rand() % 5000 == 0) show_preds(strue, t);
		}
	}
}
